DietPi-Software | frp

nonnorm opened the pull request:
<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->
**Status**: Ready

**Reference**: https://github.com/MichaIng/DietPi/issues/4395

**Commit list/description**:
<!--
- DietPi-Config | Add "Fan control" option to "Performance Options"
![Screenshot](https://xxx.github.com/images/xxx.png)
- DietPi-Config | Add "Fan control" support for Odroid C2
- DietPi-Config | Syntax fix
-->
- Multiple | Add frp
- DietPi-Software | Add G_WHIP_MENU call
- DietPi-Software | Fix G_WHIP_MENU
- DietPi-Software | Turn on dashboards
- DietPi-Software | Move frp to advanced newtorking

nonnorm committed:
v7.3
+ Multiple | Add frp

nonnorm committed:
v7.3
+ DietPi-Software | Add G_WHIP_MENU call

nonnorm committed:
v7.3
+ DietPi-Software | Fix G_WHIP_MENU

nonnorm committed:
v7.3
+ DietPi-Software | Fix mismatched quotes

nonnorm committed:
v7.3
+ DietPi-Software | Fix Download_Install directory

nonnorm committed:
v7.3
+ DietPi-Software | Fix cd into frp directory

nonnorm committed:
v7.3
+ DietPi-Software | Remove frp directory

nonnorm committed:
+ README | Add frp to 3rd party list

nonnorm committed:
v7.3
+ DietPi-Software | Turn on dashboards

nonnorm committed:
v7.3
+ DietPi-Software | Use G_EXEC on the cd

nonnorm committed:
v7.3
+ DietPi-Software | Make Admin UI accessable

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -4571,6 +4579,102 @@ amvdec_vp9' > /etc/modules-load.d/dietpi-c4-kodi.conf
+
+dashboard_port = 7500
+dashboard_user = admin
+dashboard_pwd = $GLOBAL_PW
```
Is it possible to use basic HTTP authentication here or so, at least storing credentials/password in a dedicated file that can be better secured? I try to not store the global software password in plain text, where possible, at least not world-readable. If all does not work, at least we can change the config file permissions so that it's writable only by root, readable only be frp and others have no read permissions either.

nonnorm replied:
It actually does use Basic Auth, but I think we'll need to make the file only root (and frp) readable.

MichaIng replied:
Ah, yes I meant to have the password stored in a hashed way like in `.htpasswd`.

nonnorm replied:
Maybe I should clarify: While it does let you authenticate using Basic Authentication, it hashes and encodes the password itself.
> Ah, yes I meant to have the password stored in a hashed way like in .htpasswd.

I don't quite get what you're saying here.

MichaIng replied:
Let me try to clarify:
- With the current implementation, an evil user could do `cat /etc/frp/frps.ini` to see the global software password in plain text, which is the default for many other software titles and, if not changed by the user, even the root and dietpi UNIX users login passwords. This should be prevented, when possible.
- Best would be to store the password in `/etc/frp/frps.ini` directly in a hashed way. So that the config only contains the hash while the actual password is not stored anywhere. Many software titles support this, but not all.
- Second best option would be to store it (probably together with the `dashboard_user`) in a separate authentication file, which can be specifically hardened. But I guess that is not possible, also it makes things more complicated.
- Third best option is to harden permissions of `/etc/frp/frps.ini` itself: `chmod 0600 /etc/frp/frps.ini`
    This limit read access to the service user (currently root) only, which can be seen as downside, when it's about reviewing less sensitiv config values.

nonnorm replied:
Ah, just found https://github.com/fatedier/frp#split-configures-into-different-files (how did I miss that?).  Turns out you can split it into separate files!  However, it does say only proxies, so I'll test it.
**EDIT**: Doesn't work using `[common]` on both files.  I might open a PR to implement using a hashed password, but we'll have to go with option 3 for now.

nonnorm replied:
PR up: https://github.com/fatedier/frp/pull/2448.

MichaIng committed:
v7.3 (#4442)

+ META | Update frp branch

StephanStS commented on dietpi/dietpi-software:
```diff
@@ -1305,6 +1305,14 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
+		aSOFTWARE_INTERACTIVE[$software_id]=1
+		aSOFTWARE_CATX[$software_id]=11
+		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/system_security/#frp'
 
```
Should it not be better localized in [Advanced Networking](https://dietpi.com/docs/software/advanced_networking), i.e. together with HAProxy?

MichaIng replied:
I agree. It focuses on filtering/firewalling features, but every proxy and webserver can do this, when configured accordingly. A proxy on it's own is never a security feature, but in the first place it openes another access way and can only make this additional path secure.

nonnorm committed:
v7.3
+ DietPi-Software | Move frp to advanced newtorking

nonnorm committed:
Merge branch 'frp' of https://github.com/MichaIng/DietPi into frp

nonnorm committed:
v7.3
+ DietPi-Software | Don't allow any user to access admin password

nonnorm committed:
v7.3
+ DietPi-Software | Fix uninstall

nonnorm committed:
v7.3
+ DietPi-Services | Move services to Advanced Networking

nonnorm committed:
v7.3
+ DietPi-Software | Update fallback URL

MichaIng committed:
v7.3

+ DietPi-Software | frp: Do not fail if /etc/frp does already exist (mkdir "-p" option) or if /usr/local/bin does not yet exist
+ DietPi-Software | frp: Store setup mode in a dedicated variable to avoid that subsequent G_WHIP prompts do not override the choice, e.g. such from G_EXEC error prompts
+ DietPi-Software | frp: Align systemd units, to start after network is online and start client after server (if systemd-controlled and running on the same system)
+ DietPi-Software | frp: Do not override existing configurations
+ DietPi-Software | frp: Remove downloaded directory after files have been installed
+ DietPi-Software | frp: Error-handle a few more commands, including all uninstall steps (but allow to continue uninstall in case of failure)

MichaIng committed:
v7.3

+ Tiny

nonnorm committed:
v7.3
+ DietPi-Software | Add authentication token

nonnorm committed:
v7.3
+ DietPi-Software | Fix usage of server_port

MichaIng committed:
Merge branch 'dev' into frp

MichaIng committed:
v7.3

+ CHANGELOG | frp: This proxy system has been added as install option with ID 171.

MichaIng committed:
v7.3

+ DietPi-Survey_report | Typo
